pkg/codegen/docs/gen.go

```diff
@@ -486,6 +486,40 @@ func (mod *modContext) withDocGenContext(dctx *docGenContext) *modContext {
 	return &newctx
 }
 
+func (dctx *docGenContext) getSupportedLanguages(isOverlay bool, overlaySupportedLanguages []string) []string {
+	// By default, all languages are supported.
+	if !isOverlay || len(overlaySupportedLanguages) == 0 {
+		return dctx.supportedLanguages
+	}
+	var supportedLanguages []string
+	allLanguages := codegen.NewStringSet(dctx.supportedLanguages...)
+	// Ensure that the overlay languages are a subset of the supported languages.
+	for _, lang := range overlaySupportedLanguages {
+		if allLanguages.Has(lang) {
+			supportedLanguages = append(supportedLanguages, lang)
+		}
+	}
+
+	return supportedLanguages
+}
+
+// getSupportedSnippetLanguages returns a string containing the supported snippet languages for the given type.
+// If the type is not an overlay or if there are no overlay supported languages, all languages are supported.
+// The returned string contains the supported languages joined by commas.
+func (dctx *docGenContext) getSupportedSnippetLanguages(isOverlay bool, overlaySupportedLanguages []string) string {
+	supportedLanguages := dctx.getSupportedLanguages(isOverlay, overlaySupportedLanguages)
+	supportedSnippetLanguages := make([]string, 0, len(supportedLanguages))
+	for _, lang := range supportedLanguages {
+		// Snippet languages expect "typescript" instead of "nodejs".
+		if lang == "nodejs" {
+			lang = "typescript"
+		}
+		supportedSnippetLanguages = append(supportedSnippetLanguages, lang)
+	}
+
+	return strings.Join(supportedSnippetLanguages, ",")
+}
```

Comment:
Nitpicks:

* It feels like this function exists only to workaround the `nodejs` <-> `typescript` difference. If so, it might be worth calling that out in its doc comment (and again, you could then drop the inline comment)
* If you're going to preallocate an array, I would just take the index in the loop and mutate it, no? As in `for i, lang` ... `supportedSnippetLanguages[i] = "typescript"`?
* Equally, one _could_ argue that since we know `getSupportedLanguages` will allocate a new array, you just loop over that and mutate it, since it will be GC'd when you call `strings.Join` anyway.

Reply:
`getSupportedLanguages` doesn't always allocate a new array (e.g. in the default case) and I don't think we should take that for granted. But I'll implement the first two suggestions :)